File: Easy/Parenthesis Checker/parenthesis-checker.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class Driverclass
{
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        
        //Reading total number of testcases
        int t= sc.nextInt();
        
        while(t-- >0)
        {
            //reading the string
            String st = sc.next();
            
            //calling ispar method of Paranthesis class 
            //and printing "balanced" if it returns true
            //else printing "not balanced"
            if(new Solution().ispar(st) == true)
                System.out.println("balanced");
            else
                System.out.println("not balanced");
        
        }
    }
}
// } Driver Code Ends



class Solution
{
    //Function to check if brackets are balanced or not.
    static boolean ispar(String x)
    {
        // add your code here
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put('}','{');
        mapping.put(')','(');
        mapping.put(']','[');
        for(int i=0; i<x.length(); i++){
            char c = x.charAt(i);
            if (mapping.containsKey(c)){
                char topElement = stack.empty() ? '#' : stack.pop();
                if(topElement != mapping.get(c)){
                    return false;
                }
            }else{
                stack.push(c);
            }
        }
        return stack.empty();
    }
}

File: Easy/Parenthesis Checker/parenthesis-checker.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution
{
    public:
    //Function to check if brackets are balanced or not.
    bool ispar(string s)
    {  stack<char>st;

        for(int i =0 ; i < s.length();i++){
            char ch = s[i];

            //opening bracket 

            if(ch == '(' || ch == '{'|| ch== '['){
                st.push(ch);
            }else{
                //closing bracket
                 if(!st.empty()){
                    char topCh = st.top(); 
                    if(ch == ')' and topCh == '('){
                        st.pop();
                    }
                     else if(ch == '}' and topCh == '{'){
                        st.pop();
                    }
                    else if(ch == ']' and topCh == '['){
                        st.pop();
                    }else{
                        //brackets not matching
                        return false;
                    }
                 }else{
                    return false;
                 }

                 
            }
        }
        if(st.empty()){
            return true;
        }else{
            return false;
        }
    }

};

//{ Driver Code Starts.

int main()
{
   int t;
   string a;
   cin>>t;
   while(t--)
   {
       cin>>a;
       Solution obj;
       if(obj.ispar(a))
        cout<<"balanced"<<endl;
       else
        cout<<"not balanced"<<endl;
   }
}
// } Driver Code Ends